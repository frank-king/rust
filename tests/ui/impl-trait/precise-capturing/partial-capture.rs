// See #130043 and #130031

fn main() {
    let mut data = [1, 2, 3];
    let mut i = indices(&data);
    data = [4, 5, 6];
    i.next();

    let mut i = enumerated_opaque(&data);
    i.next();

    let mut i = enumerated(&data);
    i.next();

    let mut i = enumerated_lt(&data);
    i.next();

    let mut i = enumerated_arr(data);
    i.next();
}

// No lifetime or type params captured
fn indices<T>(slice: &[T]) -> impl Iterator<Item = usize> + use<> {
    0..slice.len()
}

// `'_` and `T` are captured
fn enumerated_opaque<T>(slice: &[T]) -> impl Iterator + use<> {
    slice.iter().enumerate()
    //~^ ERROR hidden type for `impl Iterator` captures lifetime that does not appear in bounds
}

// `'_` and `T` are captured
fn enumerated_opaque_lt<T>(slice: &[T]) -> impl Iterator + use<'_> {
    //~^ ERROR `impl Trait` must mention all used type parameters in scope in `use<...>`
    slice.iter().enumerate()
}

// `'_` and `T` are captured
fn enumerated<T>(slice: &[T]) -> impl Iterator<Item = (usize, &T)> + use<> {
    //~^ ERROR `impl Trait` must mention all used type parameters in scope in `use<...>`
    slice.iter().enumerate()
}

// `'_` and `T` are captured
fn enumerated_lt<T>(slice: &[T]) -> impl Iterator<Item = (usize, &T)> + use<'_> {
    //~^ ERROR `impl Trait` must mention all used type parameters in scope in `use<...>`
    slice.iter().enumerate()
}

// `T` and `N` are captured
fn enumerated_arr<T, const N: usize>(arr: [T; N]) -> impl Iterator<Item = (usize, T)> + use<> {
    //~^ ERROR `impl Trait` must mention all used type parameters in scope in `use<...>`
    //~| ERROR `impl Trait` must mention all used const parameters in scope in `use<...>`
    <[T; N]>::into_iter(arr).enumerate()
}
