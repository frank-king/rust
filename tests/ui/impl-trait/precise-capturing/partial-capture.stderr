error[E0700]: hidden type for `impl Iterator` captures lifetime that does not appear in bounds
  --> $DIR/partial-capture.rs:29:5
   |
LL | fn enumerated_opaque<T>(slice: &[T]) -> impl Iterator + use<> {
   |                                ----     --------------------- opaque type defined here
   |                                |
   |                                hidden type `Enumerate<std::slice::Iter<'_, T>>` captures the anonymous lifetime defined here
LL |     slice.iter().enumerate()
   |     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: add `'_` to the `use<...>` bound to explicitly capture it
   |
LL | fn enumerated_opaque<T>(slice: &[T]) -> impl Iterator + use<'_> {
   |                                                             ++

error: `impl Trait` must mention all used type parameters in scope in `use<...>`
  --> $DIR/partial-capture.rs:34:44
   |
LL | fn enumerated_opaque_lt<T>(slice: &[T]) -> impl Iterator + use<'_> {
   |                         -                  ^^^^^^^^^^^^^^^^^^^^^^^
   |                         |
   |                         type parameter is implicitly captured by this `impl Trait`

error: `impl Trait` must mention all used type parameters in scope in `use<...>`
  --> $DIR/partial-capture.rs:40:34
   |
LL | fn enumerated<T>(slice: &[T]) -> impl Iterator<Item = (usize, &T)> + use<> {
   |               -                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |
   |               type parameter is implicitly captured by this `impl Trait`

error: `impl Trait` must mention all used type parameters in scope in `use<...>`
  --> $DIR/partial-capture.rs:46:37
   |
LL | fn enumerated_lt<T>(slice: &[T]) -> impl Iterator<Item = (usize, &T)> + use<'_> {
   |                  -                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                  |
   |                  type parameter is implicitly captured by this `impl Trait`

error: `impl Trait` must mention all used type parameters in scope in `use<...>`
  --> $DIR/partial-capture.rs:52:54
   |
LL | fn enumerated_arr<T, const N: usize>(arr: [T; N]) -> impl Iterator<Item = (usize, T)> + use<> {
   |                   -                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                   |
   |                   type parameter is implicitly captured by this `impl Trait`

error: `impl Trait` must mention all used const parameters in scope in `use<...>`
  --> $DIR/partial-capture.rs:52:54
   |
LL | fn enumerated_arr<T, const N: usize>(arr: [T; N]) -> impl Iterator<Item = (usize, T)> + use<> {
   |                      --------------                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      const parameter is implicitly captured by this `impl Trait`

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0700`.
